xds/internal/resolver/serviceconfig.go

```diff
@@ -53,28 +57,127 @@ type cdsBalancerConfig struct {
 	Cluster string `json:"cluster"`
 }
 
-func serviceUpdateToJSON(su xdsclient.ServiceUpdate) (string, error) {
+type route struct {
+	Path     *string                    `json:"path,omitempty"`
+	Prefix   *string                    `json:"prefix,omitempty"`
+	Regex    *string                    `json:"regex,omitempty"`
+	Headers  []*xdsclient.HeaderMatcher `json:"headers,omitempty"`
+	Fraction *wrapperspb.UInt32Value    `json:"matchFraction,omitempty"`
+	Action   string                     `json:"action"`
+}
+
+type xdsActionConfig struct {
+	ChildPolicy balancerConfig `json:"childPolicy"`
+}
+
+type xdsRoutingBalancerConfig struct {
+	Action map[string]xdsActionConfig `json:"action"`
+	Route  []*route                   `json:"route"`
+}
+
+func (r *xdsResolver) routesToJSON(routes []*xdsclient.Route) (string, error) {
+	r.updateActions(newActionsFromRoutes(routes))
+
+	// Generate routes.
+	var rts []*route
+	for _, rt := range routes {
+		t := &route{
+			Path:    rt.Path,
+			Prefix:  rt.Prefix,
+			Regex:   rt.Regex,
+			Headers: rt.Headers,
+		}
+
+		if f := rt.Fraction; f != nil {
+			t.Fraction = &wrapperspb.UInt32Value{Value: *f}
+		}
+
+		// Hash the clusters from the action, and find the assigned action
+		// names. The assigned action names are kept in r.actions, with the
+		// clusters name hash as map key.
+		//
+		// The action name is not simply the hash. For example, the hash can be
+		// "A40_B60_", but the assigned name can be "A_B_0". It's this way so
+		// the action can be reused if only weights are changing.
+		var clusterNames []string
+		clusters := rt.Action
+		for n := range clusters {
+			clusterNames = append(clusterNames, n)
+		}
+		// Hash cluster names. Sort names to be consistent.
```

Comment:
This code block of grabbing the cluster names from the action, sorting them and figuring out the `clustersOnly` or `clustersWithWeight` values is repeated at two places. Would be nicer to put them away in a helper method so that the main logical flow in this block of code (and `newActionsFromRoutes`) is easier to follow.

Reply:
Done. Moved to a function in `_action.go`.